### DISCUSSION TITLE
k8s operator taints, tolerations and affinities

### DISCUSSION THREAD

**Discussion by PaulRudin** [v6d-io/v6d#1336] [General]
I've deployed the vineyard operator to a test k8s cluster. The cluster has various node pools with a variety of taints.

How can I create a cluster so that vineyard instances will tolerate certain taints. Or how can I specify affinities to encourage vineyard instances onto certain nodes?

TIA

**Comment by dashanji** [comment]
@PaulRudin Hi, Paul. Currently,  configuring affinity and anti-affinity is not supportable refer to the [vineyard operator doc](https://v6d.io/notes/cloud-native/vineyard-operator.html#create-a-vineyard-cluster), but we will implement it ASAP.

Also, you can dump the yaml files of vineyard cluster and add some affinity or anti-affinity manually to work around it. All templates about vineyard instances are under the two directory [vineyardd](https://github.com/v6d-io/v6d/tree/main/k8s/pkg/templates/vineyardd) and [etcd](https://github.com/v6d-io/v6d/tree/main/k8s/pkg/templates/etcd). FYI.

If you encounter any problems, please feel free to report them here.

**Reply by PaulRudin** [reply]
I deployed the operator using the helm chart and that looks fine. I then applied:
```
apiVersion: k8s.v6d.io/v1alpha1
kind: Vineyardd
metadata:
  name: vineyardd-sample
  namespace: vineyard-system
```
Which creates a vineyard resource but:
```
 $ k -n vineyard-system describe vineyardds.k8s.v6d.io vineyardd-sample 
Name:         vineyardd-sample
Namespace:    vineyard-system
Labels:       <none>
Annotations:  <none>
API Version:  k8s.v6d.io/v1alpha1
Kind:         Vineyardd
Metadata:
  Creation Timestamp:  2023-04-23T13:46:27Z
  Generation:          1
  Resource Version:    1018962024
  UID:                 8f5846e2-d863-4b3b-b5df-b4fe2112718b
Spec:
  Etcd:
    Replicas:  3
  Metric Config:
    Enable:             false
    Image:              vineyardcloudnative/vineyard-grok-exporter:latest
    Image Pull Policy:  IfNotPresent
  Plugin Config:
    Backup Image:                ghcr.io/v6d-io/v6d/backup-job
    Dask Repartition Image:      ghcr.io/v6d-io/v6d/dask-repartition
    Distributed Assembly Image:  ghcr.io/v6d-io/v6d/distributed-assembly
    Local Assembly Image:        ghcr.io/v6d-io/v6d/local-assembly
    Recover Image:               ghcr.io/v6d-io/v6d/recover-job
  Replicas:                      3
  Service:
    Port:      9600
    Selector:  rpc.vineyardd.v6d.io/rpc=vineyard-rpc
    Type:      ClusterIP
  Vineyard Config:
    Etcd Endpoint:      http://etcd-for-vineyard:2379
    Etcd Prefix:        /vineyard
    Image:              vineyardcloudnative/vineyardd:latest
    Image Pull Policy:  IfNotPresent
    Size:               256Mi
    Socket:             /var/run/vineyard-kubernetes/{{.Namespace}}/{{.Name}}
    Stream Threshold:   80
    Sync CR Ds:         true
  Volume:
    Mount Path:  
    Pvc Name:    
Events:
  Type     Reason                    Age                   From                  Message
  ----     ------                    ----                  ----                  -------
  Warning  failed to apply resource  8m24s (x18 over 19m)  vineyardd-controller  encountered err: failed to create: Deployment.apps "vineyardd-sample" is invalid: [spec.template.spec.volumes[0].hostPath.path: Required value, spec.template.spec.containers[0].image: Required value, spec.template.spec.containers[0].volumeMounts[0].name: Not found: "vineyard-socket"]
```

If I understand correctly this means that the operator has tried to create a deployment for a vineyard instances, but the spec for that deployment is invalid?

**Reply by dashanji** [reply]
> If I understand correctly this means that the operator has tried to create a deployment for a vineyard instances, but the spec for that deployment is invalid?

Yeah, you are right. I'm sorry we did not update the helm chart in time, and you can get the latest helm chart now. It should be right.

Please delete the previous resources by `helm uninstall` and then run `helm repo update` to get the latest one. If you have also installed cert-manager, please delete it, as cert-manager exists as a subchart in the chart of the vineyard operator. Finally, you can repeat the previous steps again.

Thanks for reporting this!

**Reply by PaulRudin** [reply]
OK - that seems to work fine, ty.

One other question: how should I organize memory requests and limits in the context of shared memory? Suppose I have a vineyard pod, and two other pods. one which puts a 2 Gig numpy array into vineyard via ipc, and another that gets the same array from vineyard via ipc. If the memory is shared the aggregate memory used should only be 2 Gig, but from k8s perspective what will the memory use of the 3 pods look like? So what do I need to put as memory requests and limits?

**Reply by dashanji** [reply]
It should be noted that each vineyard pod is deployed on different nodes, and there are two cases:

- Stored as a LocalObject, suppose that all 2 Gig numpy arrays are stored in the vineyard pod 1, when the client on the vineyard pod 2 wants to obtain the corresponding data, since there is no data locally, the data migration operation needs to be performed first, so that the 2 Gig numpy array will be transferred to vineyard pod 2. The memory usage of object is: vineyard pod 1 -> 0, vineyard pod2 -> 2G, vineyard pod 3 -> 0

- Stored as a GlobalObject, 2 Gig numpy array is evenly distributed to 3 vineyard pods. The memory usage is: vineyard pod 1 -> 2G/3, vineyard pod2 -> 2G/3, vineyard pod 3 -> 2G/3

Currently, memory and cpu requests and limits are consistent and can be set through the following fields.
```yaml
apiVersion: k8s.v6d.io/v1alpha1
kind: Vineyardd
metadata:
  name: vineyardd-sample
  namespace: vineyard-system
spec:
  vineyard:
    memory: 4Gi
    cpu: "2"
```

**Reply by PaulRudin** [reply]
Thanks for that.

I've been experimenting a bit. One thing I don't quite understand: the deployment that the operator created does not have any memory request or limit. 

This perhaps is relevant to my earlier question about shared memory. Suppose we're only talking about local objects and 3 pods scheduled on a single node. One of those pods being a vineyard pod and the other 2 being applications that are ipc clients of the vineyard pod. Client 1 creates a numpy array of 2 Gig. Obviously it must be able to allocate 2 Gig in order to do this. It then puts this to the vineyard pod. Presumably this is shared memory, so it's not necessary to allocate another 2 Gig in either process? From k8s point of view Client 1 has used 2 Gig, vineyard has used 0 bytes for vineyard objects? Now client 2 gets the numpy array from vineyard. Again, this is accessing the same shared memory so no new memory is allocated? So if the three containers were to have memory request and limits it's only Client 1 that matters since it;'s the only one to have allocated memory? This somehow seems wrong since if we were to delete Client 1 we'd still expect client 2 to be able to get the array from Vineyard... so from k8s perspective which pod is using the memory? In any scenario you'd hope that k8s doesn't double count and consider that more memory is being used in total than is actually the case or it might refuse to schedule more pods on the node when there is actually space.

I realise that the issue here might just be that I don't really understand how shared memory works in detail...

**Reply by sighingnow** [reply]
Hi @PaulRudin,

Thanks for your insightful question.

> From k8s point of view Client 1 has used 2 Gig, vineyard has used 0 bytes for vineyard objects?

That's correct. We use [`memfd`](https://man7.org/linux/man-pages/man2/memfd_create.2.html) for shared memory so the memory usage will be accounted for pod where client 1 locates.

> Now client 2 gets the numpy array from vineyard. Again, this is accessing the same shared memory so no new memory is allocated?

Actually, yes.

> This somehow seems wrong since if we were to delete Client 1 we'd still expect client 2 to be able to get the array from Vineyard

In our previous experiments, the memory usage won't be accounted for both vineyardd pod and client 2 pod if client 1 pod exits. From k8s perspective, the memory usage won't be double counted but that is true that the k8s scheduler may over schedule pod to that machine (I'm not sure if k8s scheduling considers the actual memory usage of the node or just compute the sum of pod's requests).

The root cause is that in K8s pod spec, the memory limits/requests don't apply to shared memory. A recent version of k8s introduced a feature gate `SizeMemoryBackedVolumes` which can be used to limit the size of shared memory but still failed to specify the "requests" of shared memory.

We currently choose to account the shared memory usage to the "writer" process, i.e., the process that created the objects. Does such a design work for your scenario?

**Reply by PaulRudin** [reply]
> We currently choose to account the shared memory usage to the "writer" process, i.e., the process that created the objects. Does such a design work for your scenario?

Thanks for your explanation. I think this can work fine for my use case. I have one writer that creates all the arrays (client 1 in the above scenario) and potentially many replicas of the Client 2 pod (but these only consume - never write to shared memory). As long as I can schedule several replicas of the client 2 pod on the same node then it should work fine. I've got a proof of concept deployed now and it seems to work OK. Of course there's a way to go from a test to something that could be used in production, but I'm cautiously optimistic that it could work.

**Reply by sighingnow** [reply]
There's another approach that we can count all the memory usage for the vineyardd pod. With such a deployment, there would be no physical memory page allocation even during the `put` process and the memory usage won't be accounted for both the reader and writer. With such setting, Kubernetes scheduler won't over-schedule anymore.

Such a deployment option is not exposed to Kubernetes helm charts yet, but exposing the command line flags of vineyardd (`-reserve_memory`) should be easy.

It would be a little startup-time penalty for vineyardd container (about 1 minutes for reserving 200GB memory). Without the flag, the penalty for reserving physical memory pages is amortized during putting objects to vineyard, thus such a choose should be OK in production environment.

**Reply by PaulRudin** [reply]
Yeah, I'm wondering about just making a daemonset for the vineyard instances rather than using the operator, at least for now - because in any case I want to be able to get the vineyard instances on particular nodes; and whilst global objects are interesting and I can see some potential uses for that, I can achieve my immediate objective keeping objects local to each node.

**Reply by sighingnow** [reply]
> Yeah, I'm wondering about just making a daemonset for the vineyard instances rather than using the operator, at least for now

That could work as well. See https://github.com/v6d-io/v6d/blob/main/k8s/pkg/templates/vineyardd/deployment.yaml for the pod spec template.

> and whilst global objects are interesting and I can see some potential uses for that, I can achieve my immediate objective keeping objects local to each node.

Migrate `client.get(fetch=True)` works in these releases and further optimization on the migration implementation will be applied in the next several weeks.

**Comment by sighingnow** [comment]
Kubernetes operator is in a heavy development process currently. @PaulRudin do you have the plan to introduce Vineyard to your workflows on Kubernetes?

We are willing to know feedback and feature requests from end-users.

**Reply by PaulRudin** [reply]
I'm experimenting at the moment. I have a requirement to share memory between containers on the same k8s node and vineyard looks like it might be a good way to proceed. 

I've got a proof of concept working on a single machine using docker compose so now I want to see if I can get the same thing deployed to a k8s cluster.
